src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs

```diff
@@ -13,28 +15,95 @@ public WhenKeywordRecommender()
+            if (switchLabel == null)
             {
-                return true;
+                return false;
             }
 
-            // case int i $$
-            // case int i w$$
-            if (IsIdentifierInCasePattern(context.TargetToken.Parent))
+            var expressionOrPattern = switchLabel.ChildNodes().FirstOrDefault();
+            if (expressionOrPattern == null) // Oh well. It must have been a default label.
             {
-                return true;
+                return false;
+            }
+
+            // If the last token missing, the expression is incomplete - possibly because of missing parentheses,
+            // but not necessarily. We don't want to offer 'when' in those cases. Here are some examples that illustrate this:
+            // case |
+            // case 1 + |
+            // case (1 + 1 |
+            // case new |
+
+            // Also note that if there's a missing token inside the expression, that's fine and we do offer 'when':
+            // case (1 + ) |
+
+            var lastToken = expressionOrPattern.GetLastToken(includeZeroWidth: true);
+            if (lastToken.IsMissing)
+            {
+                return false;
+            }
+
+            // Only offer if we're at the last token of the expression/pattern, not inside it.
+
+            if (lastToken == context.TargetToken)
+            {
+                return NotATypeName(expressionOrPattern);
+            }
+            else if (lastToken == context.LeftToken && expressionOrPattern is DeclarationPatternSyntax declarationPattern)
+            {
+                // case constant w|
+
+                // We might have a partially written 'when' keyword, which causes this to be parsed as a pattern.
+                // lastToken will be 'w' (LeftToken) as opposed to 'constant' (TargetToken). We handle this as a special case.
+
+                return NotATypeName(declarationPattern.Type);
             }
 
             return false;
+
+            bool NotATypeName(SyntaxNode node)
+            {
+                // Syntactically, everything works out. We're in a pretty good spot to show 'when' now.
+                // But let's not do it just yet... Consider these cases:
+                // case SyntaxNode |
+                // case SyntaxNode w|
+                // If what we have here is known to be a type, we don't want to clutter the variable name suggestion list
+                // with 'when' since we know that the resulting code would be semantically invalid.
+
+                if (node is TypeSyntax typeSyntax)
+                {
+                    return !typeSyntax.IsPotentialTypeName(context.SemanticModel, cancellationToken);
+                }
+
+                return true;
+            }
         }
 
-        private static bool IsIdentifierInCasePattern(SyntaxNode node)
+        private static T GetAncestorUntilStatement<T>(SyntaxToken token) where T : SyntaxNode
```

Comment:
Don't we already have a GetAncestor or FindAncestor call that could already be used?

Reply:
Yes but it always goes all the way up to the root, I thought that's kinda unnecessary.. but I guess it wouldn't be a performance right? so maybe this is kinda useless

Reply:
also I don't think you can have lambdas/statements inside a switch case so I guess it wouldn't be a problem

Reply:
Usually teh root is no more than 8 items high.  So this is a microoptimization that i can't imagine ever being worth it :)

Reply:
i didn't think of this that much as an optimization as just... I just want this to be as much local search as possible, I don't really care what's outside.